temba/ussd/models.py

```diff
@@ -117,8 +117,19 @@ def handle_incoming(cls, channel, urn, date, external_id, contact=None, message_
         # check if there's an initiated PUSH session
         session = cls.objects.get_initiated_push_session(contact)
 
+        created = False
         if not session:
-            session, created = cls.objects.update_or_create(external_id=external_id, defaults=defaults)
+            try:
+                session = cls.objects.select_for_update().exclude(status__in=ChannelSession.DONE)\
```

Comment:
I'm slightly apprehensive about the `select_for_update` here. We are trying to prevent a race? Generally we like depending on redis locks over db locks just because they are a bit easier to reason about, in this case I would expect a lock on the contact being the right thing. (for example when "handling" an incoming message in a flow we lock on the contact so only one thread in the system can be moving a contact forward in a flow)